### PR TITLE
add dashboard generator

### PR DESCRIPTION
**Pull request by daiksy**

### CODE REVIEW

File: commands.go [hunk 1]
```diff
@@ -196,6 +197,7 @@ var commandDocs = map[string]commandDoc{
 	"retire":   {"", "hostIds..."},
 	"monitors": {"", "[push [--dry-run | -d] [--file-path | -F <file>] [--verbose | -v] | diff [--file-path | -F <file>] | pull [--file-path | -F <file>]]"},
 	"alerts":   {"", "[list [--service | -s <service>] [--host-status | -S <file>] [--color | -c]| close <alertIds....>]"},
+	"dashboards": {"", "[generage <file> [--print | -p]]"},
```
Comment (stefafafan) [inline commands.go]: Typo: `generage` should probably be `generate`

File: dashboards.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+type configYAML struct {
+	Title       string      `yaml:"title"`
+	URLPath     string      `yaml:"url_path"`
+	ColumnCount int         `yaml:"column_count"`
+	Graphs      []*graphDef `yaml:"graphs"`
+}
+type graphDef struct {
+	HostID      string `yaml:"host_id"`
+	ServiceName string `yaml:"service_name"`
+	RollName    string `yaml:"roll_name"`
```
Comment (stefafafan) [inline dashboards.go]: I think you probably meant `Role` not `Roll` ?

File: dashboards.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+	if confColumnCount == 1 {
+		return markdown + addItem
+	}
+	return markdown + "|" + addItem
+}
+
+func generateTableHeader(confColumnCount int) string {
+	header := ""
+	if confColumnCount > 1 {
+		for i := 0; i < confColumnCount; i++ {
```
Comment (itchyny) [inline dashboards.go]: https://golang.org/pkg/strings/#Repeat

File: dashboards.go [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+`,
+			Action: doGenerateDashboards,
+			Flags: []cli.Flag{
+				cli.BoolFlag{Name: "print, p", Usage: "markdown is output in standard output."},
+			},
+		},
+	},
+}
+
+type configYAML struct {
```
Comment (itchyny) [inline dashboards.go]: [nit] graphsConfig (YAML is one of the representation of the struct)

File: dashboards.go [hunk 5]
```diff
@@ -214,11 +226,31 @@ func doGenerateDashboards(c *cli.Context) error {
 		logger.Log("error", "url_path is required in yaml.")
 		os.Exit(1)
 	}
-	if yml.ColumnCount == 0 {
-		yml.ColumnCount = 1
+	if yml.GraphType == "" {
+		yml.GraphType = "iframe"
+	}
+	if yml.GraphType != "iframe" && yml.GraphType != "image" {
+		logger.Log("error", "graph_type should 'iframe' or 'image'.")
```
Comment (stefafafan) [inline dashboards.go]: [nit] graph_type should `be` 'iframe' or 'image'.

File: dashboards.go [hunk 6]
```diff
@@ -214,11 +226,31 @@ func doGenerateDashboards(c *cli.Context) error {
...
+	}
+	if yml.Height == 0 {
+		yml.Height = 200
+	}
+	if yml.Width == 0 {
+		yml.Width = 400
+	}
+
+	if yml.HostGraphFormat != nil && yml.GraphFormat != nil {
+		logger.Log("error", "graph should 'graphs' or 'host_graphs'.")
```
Comment (stefafafan) [inline dashboards.go]: 😕  Hm... I'm not sure what you mean by,

> graph should 'graphs' or 'host_graphs'.

File: dashboards.go [hunk 7]
```diff
@@ -251,118 +283,150 @@ func doGenerateDashboards(c *cli.Context) error {
 	return nil
 }
 
-func generateMarkDown(orgName string, graphs []*graphDef, confColumnCount int) string {
+func generateHostGraphsMarkDown(orgName string, hostGraphs *hostGraphFormat, graphType string, height int, width int, client *mackerel.Client) string {
```
Comment (stefafafan) [inline dashboards.go]: [nit] The `d` in `Markdown` is _not_ capitalized.

File: dashboards_test.go [hunk 8]
```diff
@@ -23,11 +23,29 @@ func TestHostIFrameGraph(t *testing.T) {
 }
 
 func TestServiceIFrameGraph(t *testing.T) {
```
Comment (stefafafan) [inline dashboards_test.go]: [nit] In the other file `Iframe` seems to be used in function names, but here `IFrame` is used.

File: dashboards.go [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+	}
+
+	return markdown
+}
+
+func generateGraphTableHeader(confColumnCount int) string {
+	return strings.Repeat("|:-:", confColumnCount) + "|\n"
+}
+
+func appendMarkdown(markdown string, addItem string, confColumnCount int) string {
```
Comment (itchyny) [inline dashboards.go]: `confColumnCount` is not used

File: dashboards.go [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+
+	return &markdownFactory{
+		Headline:    graphs.Headline,
+		TableHeader: tableHeader,
+		BaseGraphs:  baseGraphs,
+		ColumnCount: graphs.ColumnCount,
+	}
+}
+
+func generateGraphTableHeader(confColumnCount int) string {
```
Comment (itchyny) [inline dashboards.go]: [nit] generateAlignmentLine(count int)

File: dashboards.go [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+	markdown := ""
+	if mdf.Headline != "" {
+		markdown += fmt.Sprintf("## %s\n", mdf.Headline)
+	}
+
+	markdown += mdf.TableHeader
+
+	for i, g := range mdf.BaseGraphs {
+		markdown += "|" + g.generateGraphString(orgName)
+		if i% mdf.ColumnCount >= mdf.ColumnCount-1 || i >= len(mdf.BaseGraphs)-1 {
```
Comment (itchyny) [inline dashboards.go]: [nit] gofmt

File: dashboards.go [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+)
+
+var commandDashboards = cli.Command{
+	Name: "dashboards",
+	Subcommands: []cli.Command{
+		{
+			Name:  "generate",
+			Usage: "Generate custom dashboard",
+			Description: `
+    A custom dashboard is registered from a yaml file..
```
Comment (itchyny) [inline dashboards.go]: [message] `..` -> `.`

File: dashboards.go [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+		return expressionGraph{
+			g.Query,
+			graphType,
+			g.Period,
+			height,
+			width,
+		}
+	}
+
+	logger.Log("error", "at least one between hostId, service_name and query is required.")
```
Comment (itchyny) [inline dashboards.go]: [message] `either hostId, service_name or query should be specified.`

File: dashboards.go [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+}
+
+func doGenerateDashboards(c *cli.Context) error {
+	conffile := c.GlobalString("conf")
+
+	isStdout := c.Bool("print")
+
+	argFilePath := c.Args()
+	if len(argFilePath) < 1 {
+		logger.Log("error", "at least one argumet is required.")
```
Comment (itchyny) [inline dashboards.go]: [message] `specify a yaml file`

File: dashboards.go [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+				cli.BoolFlag{Name: "print, p", Usage: "markdown is output in standard output."},
+			},
+		},
+	},
+}
+
+type graphsConfig struct {
+	Title           string             `yaml:"title"`
+	URLPath         string             `yaml:"url_path"`
+	GraphType       string             `yaml:"graph_type"`
```
Comment (itchyny) [inline dashboards.go]: [nit] Format is better IMO.

**Pull request merged**